uaclient/api/u/pro/security/vulnerabilities/_common/v1.py

```diff
@@ -0,0 +1,325 @@
+        )
+
+    def _parse_published_date(self, published_date: str):
+        format_str = "%a, %d %b %Y %H:%M:%S GMT"
+        return datetime.strptime(published_date, format_str)
+
+    def _get_published_date(self, data_url):
+        resp = readurl(url=data_url, method="HEAD")
+        return resp.headers["last-modified"]
+
+    def _is_cache_valid(self, series: str, last_published_date: str) -> bool:
+        last_published_datetime = self._parse_published_date(
+            last_published_date
+        )
+
+        try:
+            cache_published_datetime = self._parse_published_date(
+                load_file(self._get_cache_published_date_path(series))
+            )
+        except FileNotFoundError:
+            return False
+
+        return cache_published_datetime >= last_published_datetime
+
+    def _get_cache_data(self, series: str):
+        return json.loads(load_file(self._get_cache_data_path(series)))
+
+    def _get_data_url(self, series):
+        data_name = series
+
+        enabled_services_names = [
+            s.name for s in _enabled_services(self.cfg).enabled_services
+        ]
+        if FIPSEntitlement.name in enabled_services_names:
+            data_name = "fips_{}".format(series)
+        elif FIPSUpdatesEntitlement.name in enabled_services_names:
+            data_name = "fips-updates_{}".format(series)
+
+        data_file = VULNERABILITY_DATA_TMPL.format(series=data_name)
+        return urljoin(self.cfg.vulnerability_data_url_prefix, data_file)
+
+    def get(self):
+        if self.data_file:
+            return json.loads(load_file(self.data_file))
+
+        series = self.series or get_release_info().series
+        data_url = self._get_data_url(series)
+
+        last_published_date = self._get_published_date(data_url)
+
+        if self._is_cache_valid(series, last_published_date):
+            return self._get_cache_data(series)
+
+        raw_json_data = download_xz_file_from_url(data_url)
+
+        json_data = json.loads(raw_json_data.decode("utf-8"))
+        if we_are_currently_root():
+            self._save_cache(series, json_data, last_published_date)
+
+        return json_data
+
+
+def _get_source_package_from_vulnerabilities_data(
+    vulnerabilities_data: Dict[str, Any], bin_pkg_name: str
+) -> str:
+    for pkg_name, pkg_info in vulnerabilities_data.get("packages", {}).items():
+        for _, source_info in pkg_info.get("source_versions", {}).items():
+            if bin_pkg_name in source_info["binary_packages"].keys():
+                return pkg_name
+
+    return ""
+
+
+class ProManifestSourcePackage:
+    PKG_RE = re.compile(r"^(?P<pkg>[\w\-\.\+]+)(:\w+)?\s+(?P<version>.+)$")
+
+    @staticmethod
+    def valid(manifest_file: str):
+        with open(manifest_file, "r") as f:
+            for line in f.readlines():
+                if not ProManifestSourcePackage.PKG_RE.match(line):
+                    return False
+
+        return True
```

Comment:
The problem here is that we would then need to know beforehand the manifest file pattern we expect. Today, this function exist so we can search for all of the manifest formats we support until we find one that matches.

Reply:
Ah so in the future this would return something other than True/False? Should it be named something like `detect_manifest_format` then and return an enum?

Reply:
No, the idea is that we will support multiple manifest files and this is the function we will use to see if the provide manifest matches one of the supported ones. But let me know if you think we should change it

Reply:
Okay, but if all this function does is say "Yes this matches one of the supported formats but I won't tell you which one", then the function isn't very helpful, because `parse` will have to try to parse all of the formats anyway. We could accomplish the same thing directly in `parse`:
Try format 1 - error? move on to format 2 - error? move on..
If trying to parse all formats fails, then actually raise the error.

Reply:
Ah sorry, I lost context between comments, writing a new one...

Reply:
We don't need the `valid` method if we're willing to change `parse` and the calling code a bit.

Here is semi-pseudocode

on the manifest parsing class
```python
def parse():
    try:
        parse_format()
    except Exception:
        raise Unsupported
```

the calling code that tries all manifest formats
```python
        manifest_pkgs = None
        for manifest_parser_cls in self.SUPPORTED_MANIFESTS:
            try:
                manifest_pkgs = manifest_parser_cls.parse(self.manifest_file)
                break
            except Unsupported:
                pass
        if manifest_pkgs is None:
            raise UnsupportedManifestFile
```

Reply:
Indeed, I think that works. Let me update the code